Add timeout capability to NodeRecycleThread

mach6 opened the pull request:
When specified, times out the NodeRecycleThread and
shuts down the node at the given value (in seconds).

Commit:
Add timeout capability to NodeRecycleThread and 'graceful' shutdown
option in GridForceRestartDelegateServlet

* When specified, timeout the NodeRecycleThread and
shuts down the node at the given value (in seconds).
* Add shutdown logic for NodeRecycleThread
* Better loggin for SeLionRemoteProxy logs
* Improve shutdown logic to fix bugs
* NodeForceRestartServlet returns 200 OK and a message, which is checked

mach6 commented:
@sebady @sundaramrajendran thoughts on this?

sebady commented on server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java:
```diff
@@ -245,15 +245,28 @@ public synchronized void shutdownNode() {
+            }
             appendMsgToCustomLog("Started NodeRecycleThread to restart the node once all the sessions are completed.");
-            while (getTotalUsed() > 0) {
+            while ((getTotalUsed() > 0) && (expired < timeout)) {
```
I was about to point at this loop as suspect with any ill behaved tests.   Can we test for a possible hang here with a test?

mach6 replied:
Yes. You can create a test that polls the browser and keeps it open indefinitely and run some "well-behaved" tests which reach the `uniqueSessionCount` and trigger the `NodeRecycleThread` ... In this scenario, the `NodeRecycleThread` will keep looping as long as that long running "mis-behaved" test keeps polling..

The purpose of my change is to provide for a configurable behavior which will allow the `NodeRecycleThread` to exit, should something like this happen.

sebady commented on server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java:
```diff
@@ -224,37 +247,97 @@ public synchronized void shutdownNode() {
+
+            proxyLogger.fine("Started NodeRecycleThread with " + ((timeout == -1) ? "no" : "a " + timeout + " second")
+                    + " timeout.");
+            while (keepLooping(expired, timeout)) {
```
Can we simplify the timeout loop here in run? also would a timeout of 0 be better to indicate "loop forever" (i.e no timeout) instead of "-1"

mach6 replied:
yea. I can do that.

sebady commented on server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java:
```diff
@@ -224,37 +247,97 @@ public synchronized void shutdownNode() {
+            LOGGER.entering();
+            running = false;
+            proxyLogger.fine("Interrupting NodeRecycleThread.");
+            interrupt();
```
We're setting running to false and interrupting the thread here.  Isn't one sufficient?

mach6 replied:
will fix

mach6 replied:
Actually, I'm going to keep this in place. The reason for setting `running = false` is I check it in the `catch (InterruptedException e)` block of the main `run()` loop.. If `shutdown()` was not called, then something else interrupted the recycle thread, and I want a `SEVERE` message displayed on the hub console.

sebady commented on server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java:
```diff
@@ -191,31 +181,64 @@ public TestSession getNewSession(Map<String, Object> requestedCapability) {
     }
 
+    private void stopNodeRecycleThread() {
+        if (nodeRecycleThread != null) {
```
Should this test if thread is alive? I don't think  nodeRecycleThread is ever null.

mach6 replied:
will fix

Commit:
Fix #184 - SeLion Android elements are not custom elements

- Added SeLion default Android elements (UiObject, UiTextView, UiButton)
  to AndroidSeLionElementList of codegen. Previously, they were treated
  as custom elements.. which should not have been the case
- Also added UIAElement to IOSSelionElementList
- Misc formatting changes

sebady commented on server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java:
```diff
@@ -224,37 +247,97 @@ public synchronized void shutdownNode() {
+        stopNodeRecycleThread();
+        // remove the node from the hub registry, if it reported 200 OK for NodeForceRestartServlet request
+        if (responseStatusCode == HttpStatus.SC_OK) {
+            getRegistry().removeIfPresent(this);
```
Can we unregister the node in the graceful shutdown path as well?

mach6 replied:
Already do. `requestNodeShutdown` (graceful shutdown) starts the `NodeRecycleThread` which calls `forceNodeShutdown` upon exit and when not interrupted.

mach6 commented on server/src/main/resources/config/unix/SeLionConfig.json:
```diff
@@ -1,6 +1,7 @@
 {
     "restartCycle": 60000,
     "uniqueSessionCount": 50,
+    "nodeRecycleThreadWaitTimeout": 300,
```
@sebady I was unsure about this default. Do you think a 5min timeout is a reasonable default?

sebady replied:
it seemed pretty reasonable to me.
